src/snapshot_support.h

```diff
@@ -0,0 +1,280 @@
+
+ private:
+  void WriteTag(uint8_t tag);
+  void WriteRawData(const uint8_t* data, size_t length);
+
+  v8::SnapshotCreator* creator_;
+};
+
+class SnapshotReadData final : public SnapshotDataBase {
+ public:
+  enum EmptyHandleMode {
+    kAllowEmpty,
+    kRejectEmpty
+  };
+
+  // Start reading an entry. If `expected_name` is a string, it will be matched
+  // against the string provided to StartWriteEntry(), and if there is a
+  // mismatch, this function will fail. If no such check is desired, for example
+  // because the type of the next entry is dynamic, `nullptr` can be passed.
+  // The entry type as written by StartWriteEntry() is returned.
+  v8::Maybe<std::string> StartReadEntry(const char* expected_name);
+  v8::Maybe<bool> EndReadEntry();
+
+  // Read data from the snapshot. If the read function doesn't match the write
+  // function at this position in the snapshot, the function will fail and
+  // return an empty Maybe.
+  v8::Maybe<bool> ReadBool();
+  v8::Maybe<int32_t> ReadInt32();
+  v8::Maybe<int64_t> ReadInt64();
+  v8::Maybe<uint32_t> ReadUint32();
+  v8::Maybe<uint64_t> ReadUint64();
+  v8::Maybe<V8SnapshotIndex> ReadIndex();
+  v8::Maybe<std::string> ReadString();
+
+  // Read V8 objects from the snapshot, matching the writing counterparts of
+  // these functions. If `mode` is `kRejectEmpty`, reading empty handles is
+  // regarded as a failure. Pass `kAllowEmpty` if it is expected that the
+  // corresponding write call may write an empty object.
+  // These functions do not use `v8::MaybeLocal` in order to be able to
+  // distinguish between successfully reading an empty handle and unsuccessfully
+  // reading data.
+  template <typename T>
+  inline v8::Maybe<v8::Local<T>> ReadContextIndependentObject(
+      EmptyHandleMode mode = kRejectEmpty);
+  template <typename T>
+  inline v8::Maybe<v8::Local<T>> ReadObject(
+      v8::Local<v8::Context> context, EmptyHandleMode mode = kRejectEmpty);
+
+  // Ensure that the snapshot is finished at this point, and no further data
+  // is included in the snapshot.
+  v8::Maybe<bool> Finish();
+
+  struct DumpLine {
+    size_t index;
+    size_t depth;
+    std::string description;
+
+    std::string ToString() const;
+  };
+  // Return a structured description of the snapshot contents.
+  std::pair<std::vector<DumpLine>, std::vector<Error>> Dump();
+
+  inline v8::Isolate* isolate();
+  inline void set_isolate(v8::Isolate*);
+
+  explicit inline SnapshotReadData(std::vector<uint8_t>&& storage);
+
+ private:
+  bool ReadTag(uint8_t tag);
+  v8::Maybe<uint8_t> PeekTag();
+  bool ReadRawData(uint8_t* data, size_t length);
+
+  v8::Isolate* isolate_;
+};
+
+class Snapshottable {
+ public:
+  virtual ~Snapshottable() = 0;
+
+  // Subclasses are expected to override this. The default implementation only
+  // adds an error to the snapshot data.
+  virtual void Serialize(SnapshotCreateData* snapshot_data) const;
+};
+
+class ExternalReferences {
```

Comment:
I have been thinking, for static references, can we do this via the same kind of machinery we use to register module bindings? Like adding a variant similar to `NODE_MODULE_CONTEXT_AWARE_INTERNAL`, which can later be exposed to the user land somehow. So in addition to an initialize function, modules can register a register function for these static references, which will be invoked at program startup eagerly in the snapshot builder. Ideally some checks can also be added in the snapshot builder to make sure that all the foreign references created in the initializer are registered in the register function.

Reply:
> I have been thinking, for static references, can we do this via the same kind of machinery we use to register module bindings?

@joyeecheung Yes – what’s the advantage of that?

(Aside, `dlopen()` is not going to work with the current V8 snapshot API – we’d need to know about external references before they are even loaded in memory.)

> Ideally some checks can also be added in the snapshot builder to make sure that all the foreign references created later in the initializer are registered in the register function

Yeah, I guess that would be pretty straightforward. This PR only adds external references for the bindings that are loaded during bootstrap, so I guess we’d first have to extend that for all external references before we could do that?